File: 01-machine/07-inside-a-container/02-throttled-at-30-percent-cpu/golang/quota_freeze.go
// 7.2 -- Go: the runtime that sizes itself, and the version where it started
// sizing itself from the right number.
//
// WHAT THIS DEMONSTRATES
//
//	Go's scheduler multiplexes goroutines onto GOMAXPROCS OS threads. Under
//	a 1.0-CPU quota on an 8-core host, a pre-1.25 toolchain sets GOMAXPROCS
//	to 8 and cheerfully runs eight threads flat out inside a cgroup allowed
//	100ms of CPU per 100ms. Those eight threads drain the bucket in roughly
//	12.5ms of wall clock and the kernel then freezes all of them for the
//	remaining 87.5ms. Throughput is unchanged -- the quota was always the
//	ceiling -- but every latency measurement acquires an 87ms cliff.
//
//	Go 1.25 (August 2025) made the default GOMAXPROCS the minimum of the
//	logical CPU count, the affinity mask, and the cgroup CPU bandwidth
//	limit, rounding fractional limits up, never below 2, re-checked about
//	once a second. This program prints which side of that line your
//	toolchain is on rather than assuming, because the answer changes what
//	the table below means.
//
//	"fix 1" here is literally what Go 1.25 (and Uber's automaxprocs before
//	it) does for you: set GOMAXPROCS to the quota.
//
// WHAT TO LOOK FOR IN THE OUTPUT
//  1. The GOMAXPROCS line at the top versus the enforced quota. On a
//     pre-1.25 toolchain, or with GODEBUG=containermaxprocs=0, those two
//     numbers disagree, and the gap is the bug.
//  2. Row 1 vs row 2: same offered load, same quota, only GOMAXPROCS and
//     the concurrency change. Throughput is identical; the throttle ratio
//     and the heartbeat gap are not.
//  3. The heartbeat. It burns no measurable CPU and is frozen anyway,
//     because throttling dequeues every task in the cgroup.
//
// RUN
//
//	go run quota_freeze.go
//	GODEBUG=containermaxprocs=0 go run quota_freeze.go   # the old behaviour
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"fmt"
	"math"
	mrand "math/rand"
	"os"
	"runtime"
	"runtime/pprof"
	"sort"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"
)

const (
	workMS       = 40.0 // CPU cost of one request
	offeredRate  = 9.0  // req/s -> ~0.36 CPU of demand, well under the quota
	runSeconds   = 15.0
	heartbeatDur = 10 * time.Millisecond
	periodUS     = 100000
	arrivalSeed  = 20260818
)

var hashBlock = func() []byte {
	buf := make([]byte, 256*1024)
	rand.Read(buf)
	return buf
}()

// ------------------------------------------------------------- the budget

// cpuBudget is one cgroup's worth of CFS bandwidth: a bucket refilled with
// quotaUS microseconds every periodUS, and every task parked when it is
// empty. The refill goroutine deliberately never parks -- the kernel is not
// inside your cgroup.
//
// Go needs a different accounting method from the other five languages in
// this topic, and the reason is worth knowing. Elsewhere a worker can time
// its own chunk of work with a stopwatch, because a real OS thread running
// a tight loop is running for the whole of that wall time. A goroutine is
// not: it is multiplexed onto GOMAXPROCS threads and, since Go 1.14, can be
// preempted mid-loop by a signal. Charge a goroutine's wall time and you
// bill it for time it spent descheduled -- which, at GOMAXPROCS=1, inflates
// four goroutines' usage roughly fourfold and makes the fix look like it
// did nothing.
//
// So this budget accounts the way the kernel actually does: at the GROUP
// level, by sampling the whole process's CPU time (getrusage) and charging
// the delta. Enforcement stays per-task -- workers check in between chunks
// and park when the bucket is empty, the same shape as the kernel dequeuing
// a task at its next scheduling point.
type cpuBudget struct {
	quotaUS  int64
	periodUS int64

	mu        sync.Mutex
	cond      *sync.Cond
	balance   int64
	usageUS   int64
	periods   int64
	throttled int64
	frozeThis bool
	running   bool
	lastCPUUS int64
}

func processCPUMicros() int64 {
	var usage syscall.Rusage
	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &usage); err != nil {
		return 0
	}
	user := int64(usage.Utime.Sec)*1e6 + int64(usage.Utime.Usec)
	sys := int64(usage.Stime.Sec)*1e6 + int64(usage.Stime.Usec)
	return user + sys
}

func newBudget(quotaCPUs float64, periodUS int64) *cpuBudget {
	b := &cpuBudget{
		quotaUS:   int64(quotaCPUs * float64(periodUS)),
		periodUS:  periodUS,
		running:   true,
		lastCPUUS: processCPUMicros(),
	}
	b.cond = sync.NewCond(&b.mu)
	b.balance = b.quotaUS
	go b.refillLoop()
	go b.accountantLoop()
	return b
}

// accountantLoop is the cgroup's CPU accounting. It samples more often than
// the kernel's 5ms bandwidth slice, so the group can overshoot its quota by
// at most one sample -- which is also true of the real thing, and is why a
// container's usage_usec can slightly exceed its quota within a period.
func (b *cpuBudget) accountantLoop() {
	ticker := time.NewTicker(1 * time.Millisecond)
	defer ticker.Stop()
	for range ticker.C {
		b.mu.Lock()
		if !b.running {
			b.mu.Unlock()
			return
		}
		now := processCPUMicros()
		delta := now - b.lastCPUUS
		b.lastCPUUS = now
		if delta > 0 {
			b.balance -= delta
			b.usageUS += delta
			if b.balance <= 0 {
				b.frozeThis = true
			}
		}
		b.mu.Unlock()
	}
}

func (b *cpuBudget) refillLoop() {
	ticker := time.NewTicker(time.Duration(b.periodUS) * time.Microsecond)
	defer ticker.Stop()
	for range ticker.C {
		b.mu.Lock()
		if !b.running {
			b.mu.Unlock()
			return
		}
		b.periods++
		if b.frozeThis {
			b.throttled++
		}
		b.frozeThis = false
		b.balance = b.quotaUS
		b.cond.Broadcast()
		b.mu.Unlock()
	}
}

// checkpoint is a task's scheduling point: park here if the group's bucket
// is empty. Nothing is charged -- the accountant already did that.
func (b *cpuBudget) checkpoint() {
	b.mu.Lock()
	defer b.mu.Unlock()
	for b.balance <= 0 && b.running {
		b.frozeThis = true
		b.cond.Wait()
	}
}

// park waits out a freeze without consuming quota, and reports how long it
// was frozen. This is what happens to your heartbeat, your health check and
// your metrics scraper -- punished for CPU somebody else spent.
//
// Returning the duration matters: it separates "the cgroup froze me" from
// "the Go scheduler did not get round to me". Those are different problems
// with different fixes and identical symptoms.
func (b *cpuBudget) park() float64 {
	b.mu.Lock()
	defer b.mu.Unlock()
	if b.balance > 0 {
		return 0
	}
	start := time.Now()
	for b.balance <= 0 && b.running {
		b.cond.Wait()
	}
	return float64(time.Since(start).Microseconds()) / 1000.0
}

func (b *cpuBudget) stop() {
	b.mu.Lock()
	b.running = false
	b.cond.Broadcast()
	b.mu.Unlock()
}

func (b *cpuBudget) snapshot() (usage, periods, throttled int64) {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.usageUS, b.periods, b.throttled
}

// --------------------------------------------------------------- the work

// burnCPU does a fixed amount of real work: a fixed number of hash blocks,
// not "hash until the stopwatch says 40ms". Under a quota the stopwatch
// measures freezes as well as work, so a time-bounded loop would silently
// do LESS work the more it was throttled, and throughput would stop being
// comparable between rows.
func burnCPU(blocks int, b *cpuBudget) {
	digest := sha256.New()
	for i := 0; i < blocks; i++ {
		digest.Write(hashBlock)
		if i%4 == 3 && b != nil {
			b.checkpoint() // a scheduling point, like the kernel's
		}
	}
	digest.Sum(nil)
	if b != nil {
		b.checkpoint()
	}
}

// blocksForMillis measures how many hash blocks cost targetMS on THIS
// machine. Never hardcode this: an M1 core and a c6i core are not the same
// core, and a wrong constant changes what every row below means.
func blocksForMillis(targetMS float64) int {
	start := time.Now()
	digest := sha256.New()
	for i := 0; i < 64; i++ {
		digest.Write(hashBlock)
	}
	digest.Sum(nil)
	perBlockMS := float64(time.Since(start).Microseconds()) / 1000.0 / 64
	n := int(targetMS/perBlockMS + 0.5)
	if n < 1 {
		n = 1
	}
	return n
}

// ------------------------------------------------------------- one variant

type result struct {
	completed int
	reqPerS   float64
	avgCPU    float64
	periods   int64
	throttled int64
	p50, p99  float64
	hbGap     float64
	hbFrozen  float64
}

func poissonSchedule(rate, seconds float64, seed int64) []float64 {
	// Poisson, not evenly spaced. Throttling at low average utilisation is a
	// burstiness effect: the bucket is drained by demand that clumps inside
	// one 100ms window. Evenly-spaced arrivals cannot reproduce it at all,
	// which is why hand-rolled load loops so reliably fail to find it.
	rng := mrand.New(mrand.NewSource(seed))
	var out []float64
	t := 0.0
	for t < seconds {
		out = append(out, t*1000)
		t += rng.ExpFloat64() / rate
	}
	return out
}

var workBlocks int

func runVariant(gomaxprocs, concurrency int, quotaCPUs float64) result {
	previous := runtime.GOMAXPROCS(gomaxprocs)
	defer runtime.GOMAXPROCS(previous)

	budget := newBudget(quotaCPUs, periodUS)
	defer budget.stop()

	schedule := poissonSchedule(offeredRate, runSeconds, arrivalSeed)
	start := time.Now().Add(200 * time.Millisecond)

	var mu sync.Mutex
	latencies := make([]float64, 0, len(schedule))

	cursor := make(chan int, len(schedule))
	for i := range schedule {
		cursor <- i
	}
	close(cursor)

	var wg sync.WaitGroup
	for w := 0; w < concurrency; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for i := range cursor {
				due := start.Add(time.Duration(schedule[i]) * time.Millisecond)
				if wait := time.Until(due); wait > 0 {
					time.Sleep(wait)
				}
				burnCPU(workBlocks, budget)
				mu.Lock()
				latencies = append(latencies, float64(time.Since(due).Microseconds())/1000.0)
				mu.Unlock()
			}
		}()
	}

	// Heartbeat: no CPU of its own, frozen anyway.
	hbDone := make(chan [2]float64, 1)
	hbStop := make(chan struct{})
	go func() {
		ticker := time.NewTicker(heartbeatDur)
		defer ticker.Stop()
		last := time.Now()
		maxGap, maxFrozen := 0.0, 0.0
		for {
			select {
			case <-ticker.C:
				frozen := budget.park()
				if frozen > maxFrozen {
					maxFrozen = frozen
				}
				gap := float64(time.Since(last).Microseconds()) / 1000.0
				if gap > maxGap {
					maxGap = gap
				}
				last = time.Now()
			case <-hbStop:
				hbDone <- [2]float64{maxGap, maxFrozen}
				return
			}
		}
	}()

	wg.Wait()
	close(hbStop)
	hb := <-hbDone
	hbGap, hbFrozen := hb[0], hb[1]

	usage, periods, throttled := budget.snapshot()
	sort.Float64s(latencies)
	return result{
		completed: len(latencies),
		reqPerS:   float64(len(latencies)) / runSeconds,
		avgCPU:    100 * float64(usage) / (runSeconds * 1e6),
		periods:   periods,
		throttled: throttled,
		p50:       percentile(latencies, 50),
		p99:       percentile(latencies, 99),
		hbGap:     hbGap,
		hbFrozen:  hbFrozen,
	}
}

func percentile(sorted []float64, p float64) float64 {
	if len(sorted) == 0 {
		return math.NaN()
	}
	idx := int(math.Round(p / 100 * float64(len(sorted)-1)))
	return sorted[idx]
}

// ------------------------------------------------------------- environment

func readCPUMax() (float64, bool) {
	raw, err := os.ReadFile("/sys/fs/cgroup/cpu.max")
	if err != nil {
		return 0, false
	}
	fields := strings.Fields(string(raw))
	if len(fields) < 2 || fields[0] == "max" {
		return 0, false
	}
	quota, _ := strconv.ParseFloat(fields[0], 64)
	period, _ := strconv.ParseFloat(fields[1], 64)
	if period == 0 {
		return 0, false
	}
	return quota / period, true
}

func containerAwareToolchain() bool {
	// Go 1.25 is where the default GOMAXPROCS started consulting cpu.max.
	// Parse rather than assume: this is exactly the kind of claim that goes
	// stale, and the whole table below means something different either way.
	v := strings.TrimPrefix(runtime.Version(), "go")
	parts := strings.SplitN(v, ".", 3)
	if len(parts) < 2 {
		return false
	}
	major, _ := strconv.Atoi(parts[0])
	minor, _ := strconv.Atoi(parts[1])
	return major > 1 || (major == 1 && minor >= 25)
}

func main() {
	workBlocks = blocksForMillis(workMS)
	quota, haveQuota := readCPUMax()

	fmt.Println("7.2 -- throttled at 30% CPU: Go")
	fmt.Printf("  runtime                : %s %s/%s\n",
		runtime.Version(), runtime.GOOS, runtime.GOARCH)
	fmt.Printf("  runtime.NumCPU()       : %d   <- host/affinity cores\n", runtime.NumCPU())
	fmt.Printf("  runtime.GOMAXPROCS(0)  : %d   <- OS threads that may run Go code at once\n",
		runtime.GOMAXPROCS(0))
	if haveQuota {
		fmt.Printf("  quota actually enforced: %.2f CPU (cpu.max)\n", quota)
	} else {
		fmt.Println("  quota actually enforced: none (no cgroup on this host)")
	}
	if containerAwareToolchain() {
		fmt.Println("  GOMAXPROCS default     : container-aware (toolchain >= 1.25)")
		fmt.Println("                           GODEBUG=containermaxprocs=0 restores the old behaviour.")
	} else {
		fmt.Println("  GOMAXPROCS default     : NOT container-aware (toolchain < 1.25)")
		fmt.Println("                           On this toolchain GOMAXPROCS ignores cpu.max entirely.")
		fmt.Println("                           In production you would import uber-go/automaxprocs,")
		fmt.Println("                           or set GOMAXPROCS from the same place you set the limit.")
	}
	fmt.Printf("  OS threads created     : %d   (runtime/pprof threadcreate; cumulative)\n",
		pprof.Lookup("threadcreate").Count())
	fmt.Println()

	if !haveQuota {
		fmt.Println("  !! FALLBACK: no /sys/fs/cgroup on this host")
		fmt.Println("  !! This is a userspace MODEL of cpu.max, not the Linux kernel.")
		fmt.Println("  !! Real numbers come from /sys/fs/cgroup/cpu.stat inside a container.")
		fmt.Println()
	}

	fmt.Printf("  offered load: %.0f req/s x %.0fms CPU (%d hash blocks, measured) = %.2f CPU of demand\n",
		offeredRate, workMS, workBlocks, offeredRate*workMS/1000)
	fmt.Println("  quota:        1.00 CPU. The demand is comfortably under the limit.")
	fmt.Printf("  heartbeat wants a tick every %v; %.0fs per row\n", heartbeatDur, runSeconds)
	fmt.Println()

	variants := []struct {
		label       string
		gomaxprocs  int
		concurrency int
		quotaCPUs   float64
	}{
		{fmt.Sprintf("GOMAXPROCS=%d, 4 in flight, 1.0 CPU", runtime.NumCPU()), runtime.NumCPU(), 4, 1.0},
		{"fix 1: GOMAXPROCS=1 (what 1.25 does), 1.0 CPU", 1, 4, 1.0},
		{fmt.Sprintf("fix 2: GOMAXPROCS=%d, 2.0 CPU", runtime.NumCPU()), runtime.NumCPU(), 4, 2.0},
	}

	type row struct {
		cells []string
	}
	headers := []string{"variant", "n", "req/s", "avg CPU", "throttled", "ratio",
		"p50 ms", "p99 ms", "hb gap ms", "hb frozen ms"}
	rows := make([]row, 0, len(variants))
	for _, v := range variants {
		r := runVariant(v.gomaxprocs, v.concurrency, v.quotaCPUs)
		ratio := 0.0
		if r.periods > 0 {
			ratio = float64(r.throttled) / float64(r.periods)
		}
		rows = append(rows, row{[]string{
			v.label,
			strconv.Itoa(r.completed),
			fmt.Sprintf("%.1f", r.reqPerS),
			fmt.Sprintf("%.0f%%", r.avgCPU),
			fmt.Sprintf("%d/%d", r.throttled, r.periods),
			fmt.Sprintf("%.3f", ratio),
			fmt.Sprintf("%.0f", r.p50),
			fmt.Sprintf("%.0f", r.p99),
			fmt.Sprintf("%.0f", r.hbGap),
			fmt.Sprintf("%.0f", r.hbFrozen),
		}})
		fmt.Printf("  ran: %s\n", v.label)
	}

	widths := make([]int, len(headers))
	for i, h := range headers {
		widths[i] = len(h)
		for _, r := range rows {
			if len(r.cells[i]) > widths[i] {
				widths[i] = len(r.cells[i])
			}
		}
	}
	fmt.Println()
	printRow := func(cells []string) {
		parts := make([]string, len(cells))
		for i, c := range cells {
			parts[i] = c + strings.Repeat(" ", widths[i]-len(c))
		}
		fmt.Println(strings.Join(parts, "  "))
	}
	printRow(headers)
	rule := make([]string, len(headers))
	for i := range headers {
		rule[i] = strings.Repeat("-", widths[i])
	}
	fmt.Println(strings.Join(rule, "  "))
	for _, r := range rows {
		printRow(r.cells)
	}

	fmt.Println()
	fmt.Println("  Go is usually the language that does not have the problem. Here")
	fmt.Println("  it does, and for a reason worth remembering: the scheduler is")
	fmt.Println("  excellent at not wasting threads, and completely uninterested in")
	fmt.Println("  how many CPU-seconds the kernel will sell you. Those are two")
	fmt.Println("  different questions, and until 1.25 the runtime only answered one.")
	fmt.Println()
	fmt.Println("  Read the last two columns together; they are the point of this file.")
	fmt.Println("   * 'hb frozen' is time the CGROUP held the heartbeat down.")
	fmt.Println("   * 'hb gap' is total lateness, from any cause.")
	fmt.Println("  At GOMAXPROCS=1 the frozen column collapses -- one thread cannot")
	fmt.Println("  drain the bucket faster than the kernel refills it -- while the gap")
	fmt.Println("  column may not, because now the GO SCHEDULER is the queue: four")
	fmt.Println("  CPU-bound goroutines and a heartbeat are sharing one P.")
	fmt.Println()
	fmt.Println("  Two mechanisms, one symptom, different fixes. That is the whole")
	fmt.Println("  reason cpu.stat is the first file to read: it tells the two apart")
	fmt.Println("  in ten seconds, and no latency graph ever will.")
	fmt.Println()
	fmt.Println("  Throughput is identical on every row. The quota was never the")
	fmt.Println("  throughput ceiling; it was only ever the source of the freezes.")
}

File: 01-machine/07-inside-a-container/02-throttled-at-30-percent-cpu/nodejs/quota_freeze.js
/*
 * 7.2 -- Node.js: one thread by design, and a hidden pool you did not size.
 *
 * WHAT THIS DEMONSTRATES
 *   The same experiment as the Python file next door, with the one thing
 *   only Node can show: your JavaScript runs on exactly ONE thread, so the
 *   "four runnable threads drain the bucket in 25ms" failure cannot happen
 *   to your handler code by accident. It has to be built on purpose, with
 *   worker_threads, which are separate V8 isolates rather than shared-memory
 *   threads.
 *
 *   And yet this process already has several OS threads before it runs a
 *   line of your code -- the census is printed, not guessed. libuv's thread
 *   pool (UV_THREADPOOL_SIZE,
 *   default 4, where fs, DNS and crypto live), V8's platform workers, and
 *   the GC helpers. Every one of them is inside your cgroup. Under
 *   `--cpus=1.0` a burst of concurrent `fs.readFile` or `crypto.pbkdf2`
 *   calls can drain your quota and freeze the event loop, and no amount of
 *   staring at your single-threaded handler code will explain it.
 *
 * WHAT TO LOOK FOR IN THE OUTPUT
 *   1. The at-rest thread census, printed before anything else. That is the
 *      number people are surprised by.
 *   2. Row 1 vs row 2: the same offered load and the same quota, only the
 *      worker count changes, and the throttle ratio moves while the average
 *      CPU does not.
 *   3. The heartbeat max gap. The heartbeat runs in its own worker and burns
 *      no measurable CPU. It is frozen anyway, because throttling dequeues
 *      every task in the cgroup, not just the greedy ones. That is your
 *      liveness probe failing on a container that looks idle.
 *
 * RUN
 *     node quota_freeze.js
 *
 *   On macOS the quota is a userspace model and the program says so. Inside
 *   a Linux container it reads /sys/fs/cgroup/cpu.max for ground truth.
 */
'use strict';

const os = require('os');
const fs = require('fs');
const crypto = require('crypto');
const { execSync } = require('child_process');
const {
  Worker, isMainThread, parentPort, workerData, threadId,
} = require('worker_threads');

// ---------------------------------------------------------------- config

const WORK_MS = 40;          // CPU cost of one request
const OFFERED_RATE = 9;      // req/s -> ~0.36 CPU of demand, well under quota
const RUN_SECONDS = 15;
const HEARTBEAT_MS = 10;
const PERIOD_US = 100000;

// Shared budget layout, as an Int32Array over a SharedArrayBuffer.
// Atomics.wait needs an Int32Array, which is why microseconds (not
// nanoseconds) are the unit: a period's worth of nanoseconds overflows
// int32 and the accounting would silently wrap.
const B_BALANCE = 0;       // microseconds left in this period
const B_GENERATION = 1;    // bumped on every refill; what workers wait on
const B_PERIODS = 2;
const B_THROTTLED = 3;     // periods in which someone was frozen
const B_FROZE_FLAG = 4;    // did anyone freeze in the current period
const B_USAGE_LO = 5;      // usage in microseconds, split to stay in int32
const B_USAGE_HI = 6;
const B_RUNNING = 7;
const B_SLOTS = 8;

const HASH_BLOCK = crypto.randomBytes(256 * 1024);

function nowMs() {
  // performance.timeOrigin is ms since the Unix epoch and differs per
  // worker, so adding it makes timestamps comparable ACROSS threads.
  return performance.timeOrigin + performance.now();
}

function burnCpu(targetMs, view) {
  // Wall time is charged rather than thread CPU time: Node has no
  // per-thread CPU clock, and this loop never blocks, so the two are the
  // same thing here. Python's version cannot make that assumption because
  // the GIL makes a thread's wall time include time spent not running.
  let spent = 0;
  while (spent < targetMs) {
    const mark = performance.now();
    const hash = crypto.createHash('sha256');
    for (let i = 0; i < 4; i++) hash.update(HASH_BLOCK);
    hash.digest();
    const slice = performance.now() - mark;
    spent += slice;
    if (view) chargeBudget(view, slice * 1000);
  }
}

function chargeBudget(view, micros) {
  Atomics.add(view, B_USAGE_LO, Math.round(micros));
  const balance = Atomics.sub(view, B_BALANCE, Math.round(micros)) - Math.round(micros);
  if (balance > 0) return;
  Atomics.store(view, B_FROZE_FLAG, 1);
  parkUntilRefill(view);
}

function parkUntilRefill(view) {
  while (Atomics.load(view, B_BALANCE) <= 0 && Atomics.load(view, B_RUNNING) === 1) {
    const generation = Atomics.load(view, B_GENERATION);
    // Atomics.wait blocks this OS thread outright, which is exactly what
    // the kernel does to a throttled task: dequeued, not spinning.
    Atomics.wait(view, B_GENERATION, generation, PERIOD_US / 1000);
  }
}

// ------------------------------------------------------------- worker side

if (!isMainThread) {
  const view = new Int32Array(workerData.budget);
  const dueTimes = new Float64Array(workerData.schedule);
  const cursor = new Int32Array(workerData.cursor);
  const role = workerData.role;

  if (role === 'heartbeat') {
    const gaps = [];
    let last = nowMs();
    let ticks = 0;
    const timer = setInterval(() => {
      // Burns no CPU, charges nothing -- and is frozen anyway.
      parkUntilRefill(view);
      const now = nowMs();
      gaps.push(now - last);
      last = now;
      ticks += 1;
      if (Atomics.load(view, B_RUNNING) === 0) {
        clearInterval(timer);
        parentPort.postMessage({ role, ticks, maxGap: Math.max(...gaps, 0) });
      }
    }, HEARTBEAT_MS);
  } else {
    const latencies = [];
    const loop = () => {
      while (Atomics.load(view, B_RUNNING) === 1) {
        const index = Atomics.load(cursor, 0);
        if (index >= dueTimes.length) break;
        const due = dueTimes[index];
        const now = nowMs();
        if (now < due) {
          // Nothing due yet. Sleep briefly instead of spinning; a spinning
          // worker would burn the very quota this experiment is measuring.
          Atomics.wait(view, B_GENERATION, Atomics.load(view, B_GENERATION),
            Math.min(due - now, 3));
          continue;
        }
        if (Atomics.compareExchange(cursor, 0, index, index + 1) !== index) continue;
        burnCpu(WORK_MS, view);
        latencies.push(nowMs() - due);
      }
      parentPort.postMessage({ role, latencies });
    };
    loop();
  }
  return;
}

// --------------------------------------------------------------- main side

function threadCensus() {
  try {
    if (fs.existsSync('/proc/self/status')) {
      const line = fs.readFileSync('/proc/self/status', 'utf8')
        .split('\n').find((l) => l.startsWith('Threads:'));
      return Number(line.split(/\s+/)[1]);
    }
    const out = execSync(`ps -M -p ${process.pid}`, { encoding: 'utf8' });
    return out.trim().split('\n').length - 1;
  } catch {
    return -1;
  }
}

function readCpuMax() {
  try {
    const raw = fs.readFileSync('/sys/fs/cgroup/cpu.max', 'utf8').trim();
    const [quota, period] = raw.split(' ');
    return quota === 'max' ? null : Number(quota) / Number(period || 100000);
  } catch {
    return null;
  }
}

function poissonSchedule(rate, seconds, seed) {
  // Deterministic LCG: the same arrival pattern every run, so two rows of
  // the table differ by the variable under test and nothing else.
  let state = seed >>> 0;
  const next = () => {
    state = (state * 1664525 + 1013904223) >>> 0;
    return (state + 1) / 4294967297;
  };
  const times = [];
  let t = 0;
  while (t < seconds) {
    times.push(t * 1000);
    t += -Math.log(next()) / rate;   // exponential inter-arrival
  }
  return times;
}

function percentile(sorted, p) {
  if (!sorted.length) return NaN;
  return sorted[Math.min(sorted.length - 1, Math.round((p / 100) * (sorted.length - 1)))];
}

function runVariant(workers, quotaCpus, periodUs) {
  return new Promise((resolve) => {
    const quotaUs = Math.round(quotaCpus * periodUs);
    const budget = new SharedArrayBuffer(B_SLOTS * 4);
    const view = new Int32Array(budget);
    Atomics.store(view, B_BALANCE, quotaUs);
    Atomics.store(view, B_RUNNING, 1);

    const relative = poissonSchedule(OFFERED_RATE, RUN_SECONDS, 20260818);
    const startMs = nowMs() + 300;   // let workers boot before the clock starts
    const scheduleBuf = new SharedArrayBuffer(relative.length * 8);
    new Float64Array(scheduleBuf).set(relative.map((ms) => startMs + ms));
    const cursorBuf = new SharedArrayBuffer(4);

    const pool = [];
    const latencies = [];
    let heartbeat = { ticks: 0, maxGap: NaN };
    let pending = workers + 1;

    const finish = () => {
      clearInterval(refill);
      const sorted = latencies.sort((a, b) => a - b);
      const usage = Atomics.load(view, B_USAGE_LO);
      resolve({
        completed: sorted.length,
        reqPerS: sorted.length / RUN_SECONDS,
        avgCpu: (100 * usage) / (RUN_SECONDS * 1e6),
        periods: Atomics.load(view, B_PERIODS),
        throttled: Atomics.load(view, B_THROTTLED),
        p50: percentile(sorted, 50),
        p99: percentile(sorted, 99),
        hbGap: heartbeat.maxGap,
        hbTicks: heartbeat.ticks,
      });
    };

    const spawn = (role) => {
      const worker = new Worker(__filename, {
        workerData: { budget, schedule: scheduleBuf, cursor: cursorBuf, role },
      });
      worker.on('message', (msg) => {
        if (msg.role === 'heartbeat') heartbeat = msg;
        else latencies.push(...msg.latencies);
        pending -= 1;
        worker.terminate();
        if (pending === 0) finish();
      });
      pool.push(worker);
      return worker;
    };

    for (let i = 0; i < workers; i++) spawn(`worker${i}`);
    spawn('heartbeat');

    // The refiller stands in for the kernel, so it lives on the MAIN thread
    // and is never itself frozen. The kernel is not inside your cgroup.
    const refill = setInterval(() => {
      if (Atomics.load(view, B_FROZE_FLAG) === 1) Atomics.add(view, B_THROTTLED, 1);
      Atomics.store(view, B_FROZE_FLAG, 0);
      Atomics.add(view, B_PERIODS, 1);
      Atomics.store(view, B_BALANCE, quotaUs);
      Atomics.add(view, B_GENERATION, 1);
      Atomics.notify(view, B_GENERATION);
    }, periodUs / 1000);

    setTimeout(() => {
      Atomics.store(view, B_RUNNING, 0);
      Atomics.add(view, B_GENERATION, 1);
      Atomics.notify(view, B_GENERATION);
    }, (RUN_SECONDS + 1.5) * 1000);
  });
}

function pad(value, width) {
  return String(value).padEnd(width);
}

async function main() {
  const quota = readCpuMax();
  console.log('7.2 -- throttled at 30% CPU: Node.js');
  console.log(`  runtime                : node ${process.version}, V8 ${process.versions.v8}`);
  console.log(`  os.cpus().length       : ${os.cpus().length}   <- host cores, always`);
  console.log(`  os.availableParallelism(): ${os.availableParallelism()}   <- cgroup-aware since libuv 1.49.1`);
  console.log(`  UV_THREADPOOL_SIZE     : ${process.env.UV_THREADPOOL_SIZE || '(unset -> 4)'}`);
  console.log(`  quota actually enforced: ${quota ? quota.toFixed(2) + ' CPU' : 'none (no cgroup on this host)'}`);
  console.log(`  OS threads at rest     : ${threadCensus()}  <- before any worker_threads exist`);
  console.log('');
  if (quota === null) {
    console.log('  !! FALLBACK: no /sys/fs/cgroup on this host');
    console.log('  !! This is a userspace MODEL of cpu.max, not the Linux kernel.');
    console.log('  !! Real numbers come from /sys/fs/cgroup/cpu.stat inside a container.');
    console.log('');
  }
  console.log(`  offered load: ${OFFERED_RATE} req/s x ${WORK_MS}ms CPU = ${(OFFERED_RATE * WORK_MS / 1000).toFixed(2)} CPU of demand`);
  console.log('  quota:        1.00 CPU. The demand is comfortably under the limit.');
  console.log(`  heartbeat wants a tick every ${HEARTBEAT_MS}ms; ${RUN_SECONDS}s per row`);
  console.log('');

  const variants = [
    ['4 worker_threads, 1.0 CPU (baseline)', 4, 1.0, PERIOD_US],
    ['fix 1: 1 worker_thread, 1.0 CPU', 1, 1.0, PERIOD_US],
    ['fix 2: 4 worker_threads, 2.0 CPU', 4, 2.0, PERIOD_US],
  ];

  const rows = [];
  for (const [label, workers, quotaCpus, periodUs] of variants) {
    const r = await runVariant(workers, quotaCpus, periodUs);
    rows.push([label, r.completed, r.reqPerS.toFixed(1), `${r.avgCpu.toFixed(0)}%`,
      `${r.throttled}/${r.periods}`,
      (r.periods ? r.throttled / r.periods : 0).toFixed(3),
      r.p50.toFixed(0), r.p99.toFixed(0), r.hbGap.toFixed(0)]);
    console.log(`  ran: ${label}`);
  }

  const headers = ['variant', 'n', 'req/s', 'avg CPU', 'throttled', 'ratio',
    'p50 ms', 'p99 ms', 'hb gap ms'];
  const widths = headers.map((h, i) =>
    Math.max(h.length, ...rows.map((r) => String(r[i]).length)));
  console.log('');
  console.log(headers.map((h, i) => pad(h, widths[i])).join('  '));
  console.log(widths.map((w) => '-'.repeat(w)).join('  '));
  for (const row of rows) console.log(row.map((c, i) => pad(c, widths[i])).join('  '));
  console.log('');
  console.log('  Node is the strictest version of the lesson in this lab: to get');
  console.log('  four runnable threads you had to WRITE four worker_threads. But');
  console.log(`  the at-rest census above says this process already had ${threadCensus()} OS`);
  console.log('  threads, and libuv will happily run four concurrent fs or crypto');
  console.log('  calls on them. Under a 1-CPU quota those four are enough to');
  console.log('  drain the bucket and freeze the event loop, and the stack trace');
  console.log('  you go looking at will be single-threaded and blameless.');
}

main();
